File: server/session.go
// Package server 負責 WebSocket 連線管理與玩家 session。本檔為玩家 session 綁定。
package server

import (
	"database/sql"
	"strings"
	"sync"
	"time"

	"singularity_world/db"
)

// Session 代表一名玩家的連線會話，綁定 Client 與玩家實體 ID，供遊戲邏輯辨識誰在操作。
type Session struct {
	Client          *Client
	PlayerID        string
	LastTalkAt      time.Time // 上次執行 Talk 動作的時間；用於同房玩家優先 LLM，無對話行為時才觸發 NPC 間對話
	LastTalkRoom    string    // 上次 Talk 時所在房間（供 NPC 閒聊「餘音」）
	LastTalkSnippet string    // 玩家輸入節錄（供同房 NPC 對話側面呼應）
}

// SessionStore 以 PlayerID 為 key 儲存目前上線的 session；登入時寫入、斷線時移除。
type SessionStore struct {
	mu       sync.RWMutex
	sessions map[string]*Session
}

// NewSessionStore 建立空的 SessionStore。回傳 *SessionStore，無副作用。
func NewSessionStore() *SessionStore {
	return &SessionStore{
		sessions: make(map[string]*Session),
	}
}

// Set 將 playerID 與 session 綁定；若該 playerID 已有 session 則覆寫。
// 參數：playerID 為實體 ID；s 為 *Session。
// 副作用：寫入 sessions map。
func (st *SessionStore) Set(playerID string, s *Session) {
	st.mu.Lock()
	defer st.mu.Unlock()
	st.sessions[playerID] = s
}

// Get 依 playerID 取得 Session，不存在則回傳 nil。
func (st *SessionStore) Get(playerID string) *Session {
	st.mu.RLock()
	defer st.mu.RUnlock()
	return st.sessions[playerID]
}

// Remove 移除該 playerID 的 session。
func (st *SessionStore) Remove(playerID string) {
	st.mu.Lock()
	defer st.mu.Unlock()
	delete(st.sessions, playerID)
}

// AllPlayerIDs 回傳目前所有已綁定之玩家 ID，供取得觀測者座標用。
func (st *SessionStore) AllPlayerIDs() []string {
	st.mu.RLock()
	defer st.mu.RUnlock()
	ids := make([]string, 0, len(st.sessions))
	for id := range st.sessions {
		ids = append(ids, id)
	}
	return ids
}

// AllSessions 回傳目前所有 session 的快照（slice），供外部遍歷推送用。
func (st *SessionStore) AllSessions() []*Session {
	st.mu.RLock()
	defer st.mu.RUnlock()
	list := make([]*Session, 0, len(st.sessions))
	for _, s := range st.sessions {
		list = append(list, s)
	}
	return list
}

// UpdateLastTalkAt 更新該玩家上次執行 Talk 的時間；同房玩家優先 LLM 時由 handler 呼叫。
func (st *SessionStore) UpdateLastTalkAt(playerID string) {
	st.mu.Lock()
	defer st.mu.Unlock()
	if s := st.sessions[playerID]; s != nil {
		s.LastTalkAt = time.Now()
	}
}

// sanitizeTalkSnippet 節錄玩家對白供 NPC 閒聊餘音（去換行、限長）。
func sanitizeTalkSnippet(s string) string {
	s = strings.TrimSpace(s)
	if s == "" || s == "（搭話）" {
		return ""
	}
	s = strings.ReplaceAll(s, "\r", " ")
	s = strings.ReplaceAll(s, "\n", " ")
	for strings.Contains(s, "  ") {
		s = strings.ReplaceAll(s, "  ", " ")
	}
	if len(s) > 256 {
		s = s[:256]
	}
	r := []rune(s)
	if len(r) > 48 {
		s = string(r[:48]) + "…"
	}
	return s
}

// RecordPlayerTalkForRoomEcho 記錄玩家 Talk 的時間、房間與對白節錄，供稍後同房 NPC 閒聊帶過一句「餘音」。
func (st *SessionStore) RecordPlayerTalkForRoomEcho(playerID, roomID, playerInput string) {
	st.mu.Lock()
	defer st.mu.Unlock()
	s := st.sessions[playerID]
	if s == nil {
		return
	}
	s.LastTalkAt = time.Now()
	s.LastTalkRoom = roomID
	s.LastTalkSnippet = sanitizeTalkSnippet(playerInput)
}

// PlayerTalkEchoForNpcSocial 若該房內有玩家曾在本房 Talk，且已超過 NPC 閒聊冷卻（60s）、未過期（4min），回傳節錄字串供注入 prompt。
func (st *SessionStore) PlayerTalkEchoForNpcSocial(database *sql.DB, roomID string) string {
	if roomID == "" {
		return ""
	}
	st.mu.RLock()
	defer st.mu.RUnlock()
	var bestSnippet string
	var bestAt time.Time
	for _, s := range st.sessions {
		if s == nil || s.LastTalkSnippet == "" || s.LastTalkRoom != roomID {
			continue
		}
		rid, _ := db.GetEntityRoom(database, s.PlayerID)
		if rid != roomID {
			continue
		}
		age := time.Since(s.LastTalkAt)
		if age < 60*time.Second || age > 4*time.Minute {
			continue
		}
		if bestSnippet == "" || s.LastTalkAt.After(bestAt) {
			bestSnippet = s.LastTalkSnippet
			bestAt = s.LastTalkAt
		}
	}
	return bestSnippet
}

// RoomHasPlayer 回傳該房內是否有玩家連線在場（不論是否剛 Talk）。
func (st *SessionStore) RoomHasPlayer(database *sql.DB, roomID string) bool {
	st.mu.RLock()
	defer st.mu.RUnlock()
	for _, s := range st.sessions {
		rid, _ := db.GetEntityRoom(database, s.PlayerID)
		if rid == roomID {
			return true
		}
	}
	return false
}

// RoomHasPlayerWithRecentTalk 回傳該房內是否有玩家在 within 時間內執行過 Talk；有則不觸發 NPC 間 AI 對話。
func (st *SessionStore) RoomHasPlayerWithRecentTalk(database *sql.DB, roomID string, within time.Duration) bool {
	st.mu.RLock()
	defer st.mu.RUnlock()
	for _, s := range st.sessions {
		rid, _ := db.GetEntityRoom(database, s.PlayerID)
		if rid != roomID {
			continue
		}
		if !s.LastTalkAt.IsZero() && time.Since(s.LastTalkAt) < within {
			return true
		}
	}
	return false
}

File: src/server/session.rs
//! 玩家 Session 與 `SessionStore`（對齊既有 `server/session`）。

use crate::db;
use std::collections::{HashMap, HashSet};
use std::sync::RwLock;
use std::time::{Duration, Instant};
use tokio::sync::mpsc::Sender;
use uuid::Uuid;

/// 單一玩家連線會話；`outbound` 為 WebSocket 寫入端。
#[derive(Clone)]
pub struct Session {
    /// 本次連線 id（重連時與舊 session 區分）。
    pub connection_id: Uuid,
    pub player_id: String,
    pub outbound: Option<Sender<Vec<u8>>>,
    pub last_talk_at: Option<Instant>,
    pub last_talk_room: String,
    pub last_talk_snippet: String,
}

impl std::fmt::Debug for Session {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Session")
            .field("connection_id", &self.connection_id)
            .field("player_id", &self.player_id)
            .field("has_outbound", &self.outbound.is_some())
            .finish_non_exhaustive()
    }
}

impl Session {
    pub fn new_disconnected(player_id: impl Into<String>) -> Self {
        Self {
            connection_id: Uuid::nil(),
            player_id: player_id.into(),
            outbound: None,
            last_talk_at: None,
            last_talk_room: String::new(),
            last_talk_snippet: String::new(),
        }
    }

    pub fn with_outbound(player_id: impl Into<String>, connection_id: Uuid, tx: Sender<Vec<u8>>) -> Self {
        Self {
            connection_id,
            player_id: player_id.into(),
            outbound: Some(tx),
            last_talk_at: None,
            last_talk_room: String::new(),
            last_talk_snippet: String::new(),
        }
    }

    /// 傳送原始 JSON bytes；失敗時回傳 false。
    pub fn try_send_bytes(&self, msg: Vec<u8>) -> bool {
        let Some(tx) = &self.outbound else {
            return false;
        };
        tx.blocking_send(msg).is_ok()
    }
}

/// 以 `PlayerID` 為 key 的線上 session（對齊既有 `SessionStore`）。
pub struct SessionStore {
    inner: RwLock<HashMap<String, Session>>,
}

impl Default for SessionStore {
    fn default() -> Self {
        Self::new()
    }
}

impl SessionStore {
    pub fn new() -> Self {
        Self {
            inner: RwLock::new(HashMap::new()),
        }
    }

    pub fn set(&self, player_id: &str, session: Session) {
        let mut g = self.inner.write().expect("session store poisoned");
        g.insert(player_id.to_string(), session);
    }

    pub fn get(&self, player_id: &str) -> Option<Session> {
        let g = self.inner.read().expect("session store poisoned");
        g.get(player_id).cloned()
    }

    pub fn remove(&self, player_id: &str) {
        let mut g = self.inner.write().expect("session store poisoned");
        g.remove(player_id);
    }

    /// 若目前綁定的連線 id 相符才移除（對齊既有 斷線比對 `Client`）。
    pub fn remove_if_connection(&self, player_id: &str, connection_id: Uuid) {
        let mut g = self.inner.write().expect("session store poisoned");
        if let Some(s) = g.get(player_id)
            && s.connection_id == connection_id
        {
            g.remove(player_id);
        }
    }

    pub fn all_player_ids(&self) -> Vec<String> {
        let g = self.inner.read().expect("session store poisoned");
        g.keys().cloned().collect()
    }

    pub fn all_sessions(&self) -> Vec<Session> {
        let g = self.inner.read().expect("session store poisoned");
        g.values().cloned().collect()
    }

    pub fn update_last_talk_at(&self, player_id: &str) {
        let mut g = self.inner.write().expect("session store poisoned");
        if let Some(s) = g.get_mut(player_id) {
            s.last_talk_at = Some(Instant::now());
        }
    }

    pub fn record_player_talk_for_room_echo(
        &self,
        player_id: &str,
        room_id: &str,
        player_input: &str,
    ) {
        let mut g = self.inner.write().expect("session store poisoned");
        let Some(s) = g.get_mut(player_id) else {
            return;
        };
        s.last_talk_at = Some(Instant::now());
        s.last_talk_room = room_id.to_string();
        s.last_talk_snippet = sanitize_talk_snippet(player_input);
    }

    pub fn player_talk_echo_for_npc_social(&self, room_id: &str) -> String {
        if room_id.is_empty() {
            return String::new();
        }
        let g = self.inner.read().expect("session store poisoned");
        let mut best_snippet = String::new();
        let mut best_at: Option<Instant> = None;
        let want = db::canonical_location_key(room_id);
        for s in g.values() {
            if s.last_talk_snippet.is_empty() || db::canonical_location_key(&s.last_talk_room) != want {
                continue;
            }
            let Ok(rid) = db::get_entity_room(&s.player_id) else {
                continue;
            };
            if db::canonical_location_key(&rid) != want {
                continue;
            }
            let Some(at) = s.last_talk_at else {
                continue;
            };
            let age = at.elapsed();
            if age < Duration::from_secs(60) || age > Duration::from_secs(240) {
                continue;
            }
            let replace = match best_at {
                None => true,
                Some(prev) => at > prev,
            };
            if replace {
                best_snippet.clone_from(&s.last_talk_snippet);
                best_at = Some(at);
            }
        }
        best_snippet
    }

    pub fn room_has_player(&self, room_id: &str) -> bool {
        let want = db::canonical_location_key(room_id);
        let g = self.inner.read().expect("session store poisoned");
        for s in g.values() {
            let rid = db::get_entity_room(&s.player_id).unwrap_or_default();
            if db::canonical_location_key(&rid) == want {
                return true;
            }
        }
        false
    }

    pub fn room_has_player_with_recent_talk(&self, room_id: &str, within: Duration) -> bool {
        let want = db::canonical_location_key(room_id);
        let g = self.inner.read().expect("session store poisoned");
        for s in g.values() {
            let rid = db::get_entity_room(&s.player_id).unwrap_or_default();
            if db::canonical_location_key(&rid) != want {
                continue;
            }
            if let Some(at) = s.last_talk_at
                && at.elapsed() < within
            {
                return true;
            }
        }
        false
    }

    /// 目前有玩家在線且已落在某房間的 room id 集合（對齊既有 `GetPlayerRoomMap`）。
    /// 條目為 [`db::canonical_location_key`]，使世界房間 id 與 `hex:…` 在模擬迴圈比對時一致。
    #[must_use]
    pub fn player_room_ids(&self) -> HashSet<String> {
        let mut set = HashSet::new();
        let g = self.inner.read().expect("session store poisoned");
        for s in g.values() {
            let rid = db::get_entity_room(&s.player_id).unwrap_or_default();
            if rid.is_empty() {
                continue;
            }
            set.insert(db::canonical_location_key(&rid));
        }
        set
    }
}

pub fn sanitize_talk_snippet(s: &str) -> String {
    let mut s = s.trim().to_string();
    if s.is_empty() || s == "（搭話）" {
        return String::new();
    }
    s = s
        .chars()
        .map(|c| if matches!(c, '\r' | '\n') { ' ' } else { c })
        .collect();
    while s.contains("  ") {
        s = s.replace("  ", " ");
    }
    if s.len() > 256 {
        let mut end = 256.min(s.len());
        while !s.is_char_boundary(end) {
            end -= 1;
        }
        s.truncate(end);
    }
    let chars: Vec<char> = s.chars().collect();
    if chars.len() > 48 {
        let head: String = chars[..48].iter().collect();
        return head + "…";
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_empty_and_dahua() {
        assert_eq!(sanitize_talk_snippet("   "), "");
        assert_eq!(sanitize_talk_snippet("（搭話）"), "");
    }

    #[test]
    fn sanitize_newlines_and_spaces() {
        assert_eq!(sanitize_talk_snippet("a\nb\r\nc"), "a b c".to_string());
        assert_eq!(sanitize_talk_snippet("x    y"), "x y");
    }

    #[test]
    fn store_set_get_remove() {
        let st = SessionStore::new();
        st.set("p1", Session::new_disconnected("p1"));
        assert!(st.get("p1").is_some());
        st.remove("p1");
        assert!(st.get("p1").is_none());
    }
}
